packages/apollo-server-plugin-base/src/index.ts

```diff
@@ -32,6 +32,17 @@ export interface GraphQLRequestListener<TContext = Record<string, any>> {
       'document' | 'operationName' | 'operation'
     >,
   ): ValueOrPromise<void>;
+  // If this hook is defined, it is invoked immediately before GraphQL execution
+  // would take place. If its return value resolves to a non-null
+  // GraphQLResponse, that result is used instead of executing the query.
+  // Hooks from different plugins are invoked in series and the first non-null
+  // response is used.
+  responseForOperation?(
+    requestContext: WithRequired<
+      GraphQLRequestContext<TContext>,
+      'document' | 'operationName' | 'operation'
+    >,
+  ): ValueOrPromise<GraphQLResponse | null>;
   executionDidStart?(
     requestContext: WithRequired<
       GraphQLRequestContext<TContext>,
```

Comment:
Should we also extend the `WithRequired` here to indicate that `metrics` is available?